### PR TITLE
Generic `~=` operator

### PR DESCRIPTION
**Pull request by rxwei**
Make `~=` generic over `RegexComponent` so that one can use builder components, e.g. `OneOrMore(...)`, or custom-consuming regex components as patterns.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
@swift-ci please test

File: Sources/_StringProcessing/Regex/Match.swift [hunk 1]
```diff
@@ -180,20 +180,13 @@ extension BidirectionalCollection where SubSequence == Substring {
 }
 
 @available(SwiftStdlib 5.7, *)
-extension Regex {
```
Comment (rxwei) [inline Sources/_StringProcessing/Regex/Match.swift:183]: Moving this away from Match.swift as `Regex` initializers are declared in Core.swift.

File: Tests/RegexBuilderTests/AlgorithmsTests.swift [hunk 2]
```diff
@@ -104,4 +104,38 @@ class RegexConsumerTests: XCTestCase {
...
+  func testSwitches() {
+    switch "abcde" {
+    case Regex {
+      "a"
+      ZeroOrMore(.any)
+      "f"
+    }:
+      XCTFail()
+
+    case "abc":
```
Comment (natecook1000) [inline Tests/RegexBuilderTests/AlgorithmsTests.swift:117]: Is there a way to test that this is hitting the `Equatable` version instead of the `RegexComponent` version?

File: Tests/RegexBuilderTests/AlgorithmsTests.swift [hunk 3]
```diff
@@ -104,4 +104,38 @@ class RegexConsumerTests: XCTestCase {
...
+
+    case Regex {
+      "a"
+      ZeroOrMore(.any)
+      "e"
+    }:
+      break // success
+
+    case OneOrMore(.word):
+      break // success
```
Comment (natecook1000) [inline Tests/RegexBuilderTests/AlgorithmsTests.swift:135]: Should these be in two different tests? If this is  just to test compilation, no worries.

File: Tests/RegexBuilderTests/AlgorithmsTests.swift (hunk 2 above)
Comment (rxwei) [reply Tests/RegexBuilderTests/AlgorithmsTests.swift:117]: I don't think so. Any suggestions?

File: Tests/RegexBuilderTests/AlgorithmsTests.swift (hunk 3 above)
Comment (rxwei) [reply Tests/RegexBuilderTests/AlgorithmsTests.swift]: Good point. Fixed.

**Comment by rxwei** [comment]
@swift-ci please test

File: Tests/RegexBuilderTests/AlgorithmsTests.swift (hunk 2 above)
Comment (natecook1000) [reply Tests/RegexBuilderTests/AlgorithmsTests.swift]: I don't have an idea that doesn't require making both sides generic, which I don't think we want to do. :/ We can check temporarily by printing / breaking in the operator function, but there's not really a way to test it properly.

**Review by natecook1000 — APPROVED** [review]

File: Sources/_StringProcessing/Regex/Match.swift [hunk 4]
```diff
@@ -180,20 +180,13 @@ extension BidirectionalCollection where SubSequence == Substring {
...
+@_disfavoredOverload // disambiguate from equatable patterns
+public func ~=<R: RegexComponent>(regex: R, input: String) -> Bool {
+  input.wholeMatch(of: regex) != nil
 }
 
 @available(SwiftStdlib 5.7, *)
-public func ~=<Output>(regex: Regex<Output>, input: Substring) -> Bool {
-  guard let _ = try? regex.wholeMatch(in: input) else { return false }
-  return true
+@_disfavoredOverload // disambiguate from equatable patterns
```
Comment (xedin) [inline Sources/_StringProcessing/Regex/Match.swift:189]: I think it would make sense to make them both static func in `extension RegexComponent`. Why did you add `@_disfavoredOverload`?

Comment (rxwei) [reply Sources/_StringProcessing/Regex/Match.swift:189]: When there's no `@_disfavoredOverload`, some tests failed as string patterns had two ambiguous overloads.  When I moved the operator inside `extension RegexComponent`, I no longer see those errors so that's good. Is that the effect of making an operator a static method?

File: Sources/_StringProcessing/Regex/Core.swift [hunk 5]
```diff
@@ -61,6 +61,13 @@ public struct Regex<Output>: RegexComponent {
   }
 }
 
+@available(SwiftStdlib 5.7, *)
+extension Regex {
+  public init(quoting string: String) {
+    self.init(node: .quotedLiteral(string))
+  }
+}
+
```
Comment (milseman) [inline Sources/_StringProcessing/Regex/Core.swift:70]: I have a PR in flight soon that's adding doc comments, so we might be racing on this.

File: Sources/_StringProcessing/Regex/Match.swift (hunk 4 above)
Comment (xedin) [reply Sources/_StringProcessing/Regex/Match.swift]: Yes, there is some logic we have to optimize that in the solver. If you know exactly what used to fail it would be great to reduce and add it to our test suite to make sure that doesn't regress in the future.

Comment (xedin) [reply Sources/_StringProcessing/Regex/Match.swift]: In general, if possible, it's preferred to add protocol related operators into extensions of protocols since it helps to group and filter overloads in the solver.

File: Tests/RegexBuilderTests/AlgorithmsTests.swift [hunk 6]
```diff
@@ -104,4 +104,49 @@ class RegexConsumerTests: XCTestCase {
...
+
+      case Regex {
+        "a"
+        "b"
+        "c"
+      }:
+        XCTFail()
+
+      default:
+        return
```
Comment (milseman) [inline Tests/RegexBuilderTests/AlgorithmsTests.swift:130]:
```suggestion
        break
```

Comment (milseman) [reply Tests/RegexBuilderTests/AlgorithmsTests.swift:130]: Otherwise the other tests in this function won't run. Also, would be nice to have a successful test case somewhere for switches.

**Review by milseman — APPROVED** [review]

File: Sources/_StringProcessing/Regex/Match.swift (hunk 4 above)
Comment (rxwei) [reply Sources/_StringProcessing/Regex/Match.swift]:
```swift
@available(SwiftStdlib 5.7, *)
public func ~=<R: RegexComponent>(regex: R, input: String) -> Bool {
  input.wholeMatch(of: regex) != nil
}

@available(SwiftStdlib 5.7, *)
public func ~=<R: RegexComponent>(regex: R, input: Substring) -> Bool {
  input.wholeMatch(of: regex) != nil
}

func testSwitches() {
  switch "abcde" {
  case "abc": return // error: Ambiguous use of operator '~='
  }
}
```

File: Tests/RegexBuilderTests/AlgorithmsTests.swift (hunk 6 above)
Comment (rxwei) [reply Tests/RegexBuilderTests/AlgorithmsTests.swift]: Fixed.

**Comment by rxwei** [comment]
@swift-ci please test

**Review by milseman — APPROVED** [review]
LGTM

**Comment by rxwei** [comment]
@swift-ci please test

File: Sources/_StringProcessing/Regex/Match.swift (hunk 4 above)
Comment (xedin) [reply Sources/_StringProcessing/Regex/Match.swift]: Hm, tried adding this into Match.swift and instead of ambiguity I get `switch must be exhaustive` with 04/22 snapshot of 5.7 branch.

Comment (rxwei) [reply Sources/_StringProcessing/Regex/Match.swift]: Updated the example.  It's missing a `default: break`.

File: Sources/_StringProcessing/Regex/Core.swift (hunk 5 above)
Comment (rxwei) [reply Sources/_StringProcessing/Regex/Core.swift:70]: sorry you are gonna have a conflict...

File: Sources/_StringProcessing/Regex/Match.swift (hunk 4 above)
Comment (xedin) [reply Sources/_StringProcessing/Regex/Match.swift]: Added `default` and tried with 04/23 snapshot of trunk and the project built successfully, rebased the project with origin/main just in case too.
